cmd/storage-rest-server.go

```diff
@@ -0,0 +1,358 @@
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// MakeVolHandler - make a volume.
+func (s *storageRESTServer) MakeVolHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	err := s.storage.MakeVol(volume)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// ListVolsHandler - list volumes.
+func (s *storageRESTServer) ListVolsHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	infos, err := s.storage.ListVols()
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = gob.NewEncoder(w).Encode(&infos)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// StatVolHandler - stat a volume.
+func (s *storageRESTServer) StatVolHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	info, err := s.storage.StatVol(volume)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = gob.NewEncoder(w).Encode(info)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// DeleteVolumeHandler - delete a volume.
+func (s *storageRESTServer) DeleteVolHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	err := s.storage.DeleteVol(volume)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// PrepareFileHandler - fallocate() space for a file.
+func (s *storageRESTServer) PrepareFileHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+	fileSizeStr := vars[storageRESTLength]
+	fileSize, err := strconv.Atoi(fileSizeStr)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = s.storage.PrepareFile(volume, filePath, int64(fileSize))
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// AppendFileHandler - append to a file.
+func (s *storageRESTServer) AppendFileHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+
+	buf := make([]byte, r.ContentLength)
+	_, err := io.ReadFull(r.Body, buf)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = s.storage.AppendFile(volume, filePath, buf)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// StatFileHandler - stat a file.
+func (s *storageRESTServer) StatFileHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+
+	info, err := s.storage.StatFile(volume, filePath)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = gob.NewEncoder(w).Encode(info)
+	if err != nil {
+		s.writeErrorResponse(w, err)
```

Comment:
once writing to w we cannot write error again, it will throw multiple write calls error.